File: near-mock-vm/vm/context.ts
import { Base64 } from "js-base64";
import * as path from "path";


export interface VMContext {
  current_account_id: string;
  signer_account_id: string;
  signer_account_pk: string
  predecessor_account_id: string
  //Base64 encoded
  // input_str: string;
  input: string // = bs64.encode(input_str);
  block_index: number
  block_timestamp: number;
  epoch_height: number;
  account_balance: number;
  account_locked_balance: number;
  storage_usage: number;
  attached_deposit: number;
  prepaid_gas: number;
  random_seed: string 
  is_view: boolean;
  output_data_receivers: Uint8Array
}

function findContext(_path: string = ""): VMContext | null {
  let paths = [
                _path,
                path.join(process.cwd(),
                "assembly", "__tests__"),
                process.cwd(), 
                __dirname
              ]
              .map(p => {
                if (!p.endsWith("context.json")) {
                  return path.join(p, "context.json")
                }
                return p;
              });
  let _paths = paths.filter(p => {
      try {
        require.resolve(p);
      } catch {
          return false;
      }
      return true
  });
  let context: VMContext = _paths.length > 0 ? require(_paths[0]) : null;
  if (context != null) {
    console.log("found path: " + _paths[0]);
    context.input = Base64.encode(context.input);
  }
  return context;
  
}



const current_account_id = "alice"; 
const signer_account_id = "bob";
const signer_account_pk = "HuxUynD5GdrcZ5MauxJuu74sGHgS6wLfCqqhQkLWK";
const predecessor_account_id = "carol";
const input_str = "{\"a\":21,\"b\":21}";
const block_index = 10;
const block_timestamp = 42;
const epoch_height = 20;
const account_balance = 2;
const account_locked_balance = 1;
const storage_usage = 12;
const attached_deposit = 2;
const prepaid_gas = 10**(14);
const random_seed = "HuxUynD5GdrcZ5MauxJuu74sGHgS6wLfCqqhQkLWK";
const is_view = false;
const output_data_receivers= new Uint8Array([]);

export function createDefault(): VMContext {
  const _default = {
    /// The account id of the current contract that we are executing.
    current_account_id, // string
    /// The account id of that signed the original transaction that led to this
    /// execution.
    signer_account_id, // string
    /// The public key that was used to sign the original transaction that led to
    /// this execution.
    signer_account_pk, // string base58
    predecessor_account_id, // string
    input: Base64.encode(input_str), // JSON string
    block_index, // u128
    block_timestamp,
    epoch_height,
    account_balance,
    account_locked_balance,
    storage_usage,
    attached_deposit,
    prepaid_gas,
    random_seed,
    is_view,
    output_data_receivers,
  }
  return _default
}


export function createContext(_path?: string): VMContext {
  let context =  findContext(_path) || createDefault();
  context.input = Base64.encode(context.input);
  return context;
}

File: near-sdk/src/test_utils/context.rs
use crate::mock::MockedBlockchain;
use crate::test_utils::test_env::*;
use crate::{test_vm_config, AccountId};
use crate::{BlockHeight, EpochHeight, Gas, NearToken, PromiseResult, PublicKey, StorageUsage};
use near_parameters::RuntimeFeesConfig;
use near_primitives_core::config::ViewConfig;
use std::convert::TryInto;

/// Returns a pre-defined account_id from a list of 6.
pub fn accounts(id: usize) -> AccountId {
    ["alice", "bob", "charlie", "danny", "eugene", "fargo"][id].parse().unwrap()
}

/// Simple VMContext builder that allows to quickly create custom context in tests.
#[derive(Clone)]
pub struct VMContextBuilder {
    pub context: VMContext,
}

impl Default for VMContextBuilder {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Clone)]
/// Context for the contract execution.
pub struct VMContext {
    /// The account id of the current contract that we are executing.
    pub current_account_id: AccountId,
    /// The account id of that signed the original transaction that led to this
    /// execution.
    pub signer_account_id: AccountId,
    /// The public key that was used to sign the original transaction that led to
    /// this execution.
    pub signer_account_pk: PublicKey,
    /// If this execution is the result of cross-contract call or a callback then
    /// predecessor is the account that called it.
    /// If this execution is the result of direct execution of transaction then it
    /// is equal to `signer_account_id`.
    pub predecessor_account_id: AccountId,
    /// The input to the contract call.
    /// Encoded as base64 string to be able to pass input in borsh binary format.
    pub input: Vec<u8>,
    /// The current block height.
    pub block_index: BlockHeight,
    /// The current block timestamp (number of non-leap-nanoseconds since January 1, 1970 0:00:00 UTC).
    pub block_timestamp: u64,
    /// The current epoch height.
    pub epoch_height: EpochHeight,

    /// The balance attached to the given account. Excludes the `attached_deposit` that was
    /// attached to the transaction.
    pub account_balance: NearToken,
    /// The balance of locked tokens on the given account.
    pub account_locked_balance: NearToken,
    /// The account's storage usage before the contract execution
    pub storage_usage: StorageUsage,
    /// The balance that was attached to the call that will be immediately deposited before the
    /// contract execution starts.
    pub attached_deposit: NearToken,
    /// The gas attached to the call that can be used to pay for the gas fees.
    pub prepaid_gas: Gas,
    /// Initial seed for randomness
    pub random_seed: [u8; 32],
    /// If Some, it means that execution is made in a view mode and defines its configuration.
    /// View mode means that only read-only operations are allowed.
    /// See <https://nomicon.io/Proposals/0018-view-change-method.html> for more details.
    pub view_config: Option<ViewConfig>,
    /// How many `DataReceipt`'s should receive this execution result. This should be empty if
    /// this function call is a part of a batch and it is not the last action.
    pub output_data_receivers: Vec<AccountId>,
}

impl VMContext {
    pub fn is_view(&self) -> bool {
        self.view_config.is_some()
    }
}

#[allow(dead_code)]
impl VMContextBuilder {
    pub fn new() -> Self {
        Self {
            context: VMContext {
                current_account_id: alice(),
                signer_account_id: bob(),
                signer_account_pk: vec![0u8; 33].try_into().unwrap(),
                predecessor_account_id: bob(),
                input: vec![],
                block_index: 0,
                block_timestamp: 0,
                epoch_height: 0,
                account_balance: NearToken::from_yoctonear(10u128.pow(26)),
                account_locked_balance: NearToken::from_near(0),
                storage_usage: 1024 * 300,
                attached_deposit: NearToken::from_near(0),
                prepaid_gas: Gas::from_tgas(300),
                random_seed: [0u8; 32],
                view_config: None,
                output_data_receivers: vec![],
            },
        }
    }

    pub fn current_account_id(&mut self, account_id: AccountId) -> &mut Self {
        self.context.current_account_id = account_id;
        self
    }

    pub fn signer_account_id(&mut self, account_id: AccountId) -> &mut Self {
        self.context.signer_account_id = account_id;
        self
    }

    pub fn signer_account_pk(&mut self, pk: PublicKey) -> &mut Self {
        self.context.signer_account_pk = pk;
        self
    }

    pub fn predecessor_account_id(&mut self, account_id: AccountId) -> &mut Self {
        self.context.predecessor_account_id = account_id;
        self
    }

    #[deprecated(since = "4.1.2", note = "Use `block_height` method instead")]
    pub fn block_index(&mut self, block_index: BlockHeight) -> &mut Self {
        self.context.block_index = block_index;
        self
    }

    pub fn block_height(&mut self, block_height: BlockHeight) -> &mut Self {
        self.context.block_index = block_height;
        self
    }

    pub fn block_timestamp(&mut self, block_timestamp: u64) -> &mut Self {
        self.context.block_timestamp = block_timestamp;
        self
    }

    pub fn epoch_height(&mut self, epoch_height: EpochHeight) -> &mut Self {
        self.context.epoch_height = epoch_height;
        self
    }

    pub fn account_balance(&mut self, amount: NearToken) -> &mut Self {
        self.context.account_balance = amount;
        self
    }

    pub fn account_locked_balance(&mut self, amount: NearToken) -> &mut Self {
        self.context.account_locked_balance = amount;
        self
    }

    pub fn storage_usage(&mut self, usage: StorageUsage) -> &mut Self {
        self.context.storage_usage = usage;
        self
    }

    pub fn attached_deposit(&mut self, amount: NearToken) -> &mut Self {
        self.context.attached_deposit = amount;
        self
    }

    pub fn prepaid_gas(&mut self, gas: Gas) -> &mut Self {
        self.context.prepaid_gas = gas;
        self
    }

    pub fn random_seed(&mut self, seed: [u8; 32]) -> &mut Self {
        self.context.random_seed = seed;
        self
    }

    pub fn is_view(&mut self, is_view: bool) -> &mut Self {
        self.context.view_config =
            if is_view { Some(ViewConfig { max_gas_burnt: 200000000000000 }) } else { None };
        self
    }

    pub fn build(&self) -> VMContext {
        self.context.clone()
    }
}

/// Initializes the [`MockedBlockchain`] with a single promise result during execution.
#[deprecated(since = "4.0.0", note = "Use `testing_env!` macro to initialize with promise results")]
pub fn testing_env_with_promise_results(context: VMContext, promise_result: PromiseResult) {
    let storage = crate::mock::with_mocked_blockchain(|b| b.take_storage());

    //? This probably shouldn't need to replace the existing mocked blockchain altogether?
    //? Might be a good time to remove this utility function altogether
    crate::env::set_blockchain_interface(MockedBlockchain::new(
        context,
        test_vm_config(),
        RuntimeFeesConfig::test(),
        vec![promise_result],
        storage,
        Default::default(),
        None,
    ));
}
